File: solutions/fsharp/circular-buffer/1/CircularBuffer.fs
module CircularBuffer

type CircularBuffer = {
    Size: int
    Values: int list
}

let mkCircularBuffer size = { Size = size; Values = [] }

let clear (buffer: CircularBuffer) = { buffer with Values = [] }

let write value (buffer: CircularBuffer) =
    if buffer.Size = buffer.Values.Length then
        failwith "Buffer full."
    { buffer with Values = buffer.Values @ [ value ] }
        
let forceWrite value (buffer: CircularBuffer) =
    if buffer.Size <> buffer.Values.Length then
        write value buffer
    else
        {    
            buffer with
                Values = buffer.Values[1..] @ [ value ]
        }

let read (buffer: CircularBuffer) =
    match buffer.Values with
    | [] -> failwith "You need to implement this function."
    | head::tail -> (head, { buffer with Values = tail })

File: solutions/scala/circular-buffer/1/src/main/scala/CircularBuffer.scala
class EmptyBufferException extends Exception("The buffer is empty.")
class FullBufferException extends Exception("The buffer is full.")

class CircularBuffer(var capacity: Int) {
  private val buffer: Array[Option[Int]] = Array.fill(capacity)(None)
  private var readIndex = 0
  private var writeIndex = 0
  private var size = 0

  def write(value: Int): Unit = {
    if (size == capacity) throw new FullBufferException
    buffer(writeIndex) = Some(value)
    writeIndex = (writeIndex + 1) % capacity
    size += 1
  }

  def read(): Int = {
    if (size == 0) throw new EmptyBufferException
    val value = buffer(readIndex).get
    buffer(readIndex) = None
    readIndex = (readIndex + 1) % capacity
    size -= 1
    value
  }

  def overwrite(value: Int): Unit = {
  if (size < capacity) {
    buffer(writeIndex) = Some(value)
    writeIndex = (writeIndex + 1) % capacity
    size += 1
  } else {
    buffer(writeIndex) = Some(value)
    writeIndex = (writeIndex + 1) % capacity
    readIndex = (readIndex + 1) % capacity
  }
}

  def clear(): Unit = {
    for (i <- buffer.indices) buffer(i) = None
    readIndex = 0
    writeIndex = 0
    size = 0
  }
}
